### PR TITLE
ENH: allow overlay for all geometry types

### PR DESCRIPTION
**Pull request by martinfleis**
Following the discussion in #821, this PR will eventually allow overlay for all types of geometries. The current implementation is robust enough to work with minimal changes. The main block in the master is use of `buffer(0)` at some points, causing LineStrings and Points collapse into empty geometries. I think there was an intention to use it only optionally as there is `make_valid=True` argument, which is actually unused.

The only failing test is a bit corner case, we should decide what to do with. In the following example, master returns two Polygons, while this PR the same Polygons,  but also one LineString and two (identical) Points. In theory, the second case is correct, in practice it is the question if it is the intended result.

```
s2 = gpd.GeoSeries([Polygon([(1, 1), (3, 1), (3, 3), (1, 3)]),
                    Polygon([(3, 3), (5, 3), (5, 5), (3, 5)])])
df2 = gpd.GeoDataFrame({'col2': [1, 2], 'geometry': s2})

polys3 = gpd.GeoSeries([Polygon([(1, 1), (3, 1), (3, 3), (1, 3)]),
                        Polygon([(-1, 1), (1, 1), (1, 3), (-1, 3)]),
                        Polygon([(3, 3), (5, 3), (5, 5), (3, 5)])])
df3 = gpd.GeoDataFrame({'geometry': polys3, 'col3': [1, 2, 3]})

ov = gpd.overlay(df3, df2)
```
```
   col3  col2                             geometry
0     1     1  POLYGON ((3 1, 1 1, 1 3, 3 3, 3 1))
1     2     1                LINESTRING (1 1, 1 3)
2     3     1                          POINT (3 3)
3     1     2                          POINT (3 3)
4     3     2  POLYGON ((5 3, 3 3, 3 5, 5 5, 5 3))
```

Second issue might be the fact, that overlay of two geometry types often returns GeoSeries with mixed geometries. As this result is correct and GeoDataFrame has no issue with it, user needs to be aware of the fact that it needs to be separated before saving to file.

Before I'll do any further work on this, I'd like to have an agreement how it should behave in the cases above (and probably other I haven't found yet).

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Any idea how eg QGIS does this? (or do they only support polygons?)

**Comment by martinfleis** [comment]
QGIS returns only polygons. It always returns only `left` geometry type. In a sense, we can offer more precise results here. Maybe we can have parameter for `strict` geometry type (returns only geometry type of left gdf) and `full` (returns everything). Because intersection of two polygons sharing an edge is a line and I we should be able to return it.

**Comment by martinfleis** [comment]
Using `strict=True`, overlay now checks geometry type of left gdf and returns only matching types. The same behaviour as QGIS has.  While using `strict=False`, it returns complete result, having mixed geometries. (except `difference` at the moment)

```
s2 = gpd.GeoSeries([Polygon([(1, 1), (3, 1), (3, 3), (1, 3)]),
                    Polygon([(3, 3), (5, 3), (5, 5), (3, 5)])])
df2 = gpd.GeoDataFrame({'col2': [1, 2], 'geometry': s2})

polys3 = gpd.GeoSeries([Polygon([(1, 1), (3, 1), (3, 3), (1, 3)]),
                        Polygon([(-1, 1), (1, 1), (1, 3), (-1, 3)]),
                        Polygon([(3, 3), (5, 3), (5, 5), (3, 5)])])
df3 = gpd.GeoDataFrame({'geometry': polys3, 'col3': [1, 2, 3]})
```

`gpd.overlay(df3, df2, strict=False)`
```
   col3  col2                             geometry
0     1     1  POLYGON ((1 1, 1 3, 3 3, 3 1, 1 1))
1     2     1                LINESTRING (1 3, 1 1)
2     3     1                          POINT (3 3)
3     1     2                          POINT (3 3)
4     3     2  POLYGON ((3 3, 3 5, 5 5, 5 3, 3 3))
```
`gpd.overlay(df3, df2, strict=True)`
```
   col3  col2                             geometry
0     1     1  POLYGON ((1 1, 1 3, 3 3, 3 1, 1 1))
1     3     2  POLYGON ((3 3, 3 5, 5 5, 5 3, 3 3))
```

**Comment by martinfleis** [comment]
Results of overlay on mixed geometries follow the logic of shapely's [difference](https://shapely.readthedocs.io/en/latest/manual.html#object.difference):
> Shapely can not represent the difference between an object and a lower dimensional object (such as the difference between a polygon and a line or point) as a single object, and in these cases the difference method returns a copy of the object named self. 

One might expect that LineString will cut Polygon in half, but for this reason that does not happen. See [examples](https://gist.github.com/martinfleis/ae58340afff65ffeb825665b21786b2d).

In the case of union, this behaviour is a bit unintuitive (LineStrings are cut, Polygons not), but If we properly document it as shapely did, it should be okay. The rest of the cases seems to work as one would expect (with `strict` keyword). What do you think?

**Comment by martinfleis** [comment]
Alright, this is finally ready. Summary:

This PR enables all geometry types in `overlay`, not only (Multi)Polygons. It introduces new keyword `strict` controlling the behaviour - if set to True (default), overlay returns only geometries of the same type as input df1 has. If set to False, it returns all geometries resulting from particular operations (intersection can result in LineStrings, Collections...). For example, you can now use difference on LineStrings and Polygons (for a removal of edges shared by the LineString df and Polygon df). And all the other as long as it makes sense for selected operation.

At the same time, I have cleaned overlay.py and removed the old version of overlay which was kept only for a reference if new one was faulty. As it was merged in 0.4, I assume it is no longer needed for that purpose. I have also removed `use_sindex` deprecated in 0.4. 

Tests are done in the same way as the rest of overlay, that is why there is a bunch of geojsons.

**Comment by martinfleis** [comment]
I have changed `strict` keyword to `keep_geom_type` to be consistent with clip #1128. The failing Travis is not related to this PR (just saying). Ready for review.

File: geopandas/tools/overlay.py [hunk 1]
```diff
@@ -225,7 +39,10 @@ def _overlay_intersection(df1, df2):
         left.reset_index(drop=True, inplace=True)
         right = df2.geometry.take(pairs["__idx2"].values)
         right.reset_index(drop=True, inplace=True)
-        intersections = left.intersection(right).buffer(0)
+        intersections = left.intersection(right)
+        intersections = intersections.apply(
+            lambda g: g.buffer(0) if g.type in ["Polygon", "MultiPolygon"] else g
+        )
```
Comment (jorisvandenbossche) [inline geopandas/tools/overlay.py]: We might want to write this without an apply (to benefit from vectorized operations in the future, but so can also be done later)

**Comment by martinfleis** [comment]
> But before we didn't do such a check, so how come that before we only returned polygons?

@jorisvandenbossche We did. We just didn't allow anything else. That is why this PR was not so complicated.
https://github.com/geopandas/geopandas/blob/0dde1c7253ee4755f2ab8656c3bfa74c617e8d3a/geopandas/tools/overlay.py#L373-L380

**Comment by jorisvandenbossche** [comment]
But that's on the input? I was meaning the output. 
For example, the example you showed in the top post, is with only polygons as input, but having also lines and points in the output. Now you added filtering on the output to only return polygons. But how come we had that before?

**Comment by martinfleis** [comment]
@jorisvandenbossche Because `.buffer(0)` on LineString or Point generates empty polygon which is then dropped.

https://github.com/geopandas/geopandas/blob/0dde1c7253ee4755f2ab8656c3bfa74c617e8d3a/geopandas/tools/overlay.py#L228-L232

File: geopandas/tests/test_overlay.py [hunk 2]
```diff
@@ -335,3 +327,149 @@ def test_correct_index(dfs):
...
+        result = overlay(df1, df3, how=how, keep_geom_type=keep_geom_type)
+    elif geom_types == "poly_point":
+        result = overlay(df1, df4, how=how, keep_geom_type=keep_geom_type)
+    elif geom_types == "line_poly":
+        result = overlay(df3, df1, how=how, keep_geom_type=keep_geom_type)
+    elif geom_types == "point_poly":
+        result = overlay(df4, df1, how=how, keep_geom_type=keep_geom_type)
+
+    if not _FIONA18 and result.empty:
+        raise pytest.skip()
```
[the remark this answers is not in the document]
Comment (martinfleis) [reply geopandas/tests/test_overlay.py]: It was due to different errors to catch between versions, I changed that to a more explicit version.

**Comment by martinfleis** [comment]
@jorisvandenbossche comments implemented.

**Comment by jorisvandenbossche** [comment]
> Because .buffer(0) on LineString or Point generates empty polygon which is then dropped.

Ah, that detail I missed!

**Comment by gaoag** [comment]
hi;

the 0.8.0 release seems not to have kept the implementation here. The intended behavior from a previous comment

https://github.com/geopandas/geopandas/pull/1110#issuecomment-523076424

isn't found in the 0.8.0 release. This is what happens when I do the same thing as the post above, but in 0.8.0 (this behavior exists no matter if I have keep_geom_type as True or False):

image

**Comment by martinfleis** [comment]
@gaoag are you sure you are using 0.8.0? It works as intended for me.
